bill/calculator.go

```diff
@@ -175,33 +175,36 @@ func calculate(doc billable) error {
 	if t.RetainedTax != nil {
 		t.Payable = t.Payable.Subtract(*t.RetainedTax)
 	}
-	if t.Rounding != nil {
-		// BT-144 in EN16931
-		t.Payable = t.Payable.Add(*t.Rounding)
-	}
-
 	// Remove taxes object if it doesn't contain any categories
 	if len(t.Taxes.Categories) == 0 {
 		t.Taxes = nil
 	}
 
+	// Before calculating the amount due and advances, we need to round
+	// everything. Payments reflect real monetary values and can never
+	// be fractions of the currency.
+	roundLines(doc.getLines())
+	roundDiscounts(doc.getDiscounts(), cur)
+	roundCharges(doc.getCharges(), cur)
+	t.round(zero)
+
+	if t.Rounding != nil {
+		// BT-144 in EN16931
+		t.Payable = t.Payable.Add(*t.Rounding)
+	}
 	if pd := doc.getPaymentDetails(); pd != nil {
 		pd.calculateAdvances(zero, t.Payable)
-
-		// Deal with advances, if any
+		// Deal with advances, if any. Note that in the current
+		// implementation multiple percentage advances are likely to
+		// suffer rounding errors. It usually better for users to use
+		// fixed payment amounts if possible.
```

Comment:
This is very true. And makes me wonder where GOBL should actually be making this calculation. An advance is supposed to be a payment already made and so, already calculated. I wonder if, to ensure no ambiguity can happen, GOBL should only accept an amount (not a percentage) (and only with the currency decimals)

Reply:
I think the percent calculation is useful, but it does indeed need a better warning. I'll update the advance object's description to see if that helps. 👍